File: internal/config/videoformat.go
// The videoFormatOverride grammar: one string, one parser, one canonical form.
//
// Owner: WP-1, with internal/config. It is a file of its own rather than
// eighty more lines in config.go because it is the only thing in this package
// that PARSES anything — everything else here reads a number, a port or an
// enum out of JSON and checks its range — and because the argument for the
// representation is long enough that burying it between two struct fields
// would lose it.
//
// # What this exists to decide
//
// M2L-X can be configured into any video format and every source feeding it
// must match; the format is derivable from switcher_status only while some node
// is actually streaming (internal/m2lx parses state.streams.video.format for
// exactly that), and it is derivable from nothing at all when none is.
// MEASURED against the live matchH instance, 2026-08-15: all 35 nodes reported
// "format": null with stream_state "stopped" — our own commentary input, cam4
// "COMMS", among them — and seven plausible REST paths for an instance-wide
// setting all answered 404. So a position that comes up first, which is the
// normal case, has to be TOLD, and config.Config.VideoFormatOverride is where
// it is told. This file is what turns what the operator typed into numbers a
// capsfilter can be built from.
//
// # Why a string and not a struct
//
// The full argument is on the field in config.go. In one line each: a string
// keeps an unexpected value VISIBLE instead of turning it into three plausible
// zeros — the same reason m2lx.VideoFormat carries Raw beside its parsed
// fields — and a string cannot HALF-ARRIVE through this package's
// unmarshal-onto-the-live-struct merge, where a nested {"width":1280} would
// leave the old height in place and conform to 1280x1080.
//
// # WHAT IS GIVEN UP BY THAT CHOICE, stated rather than glossed
//
// A string can be misspelled and a struct cannot. Everything below is the price
// of that, and it is paid in one place so that no caller has to think about it:
// ONE grammar, ONE parser, and config.Validate refusing a value this function
// rejects — with the field name, the value and the accepted form in the message
// — so that a typo is a sentence on the Settings screen and never a
// not-negotiated (-4) from inside GStreamer half a minute after START.
//
// The second thing given up is expressiveness, deliberately. The grammar below
// describes progressive video and nothing else. It cannot say interlaced, it
// cannot say a colour space, a bit depth or a chroma sampling, and it does not
// accept a raw n/d fraction. Each of those is refused rather than tolerated,
// because a format this application accepts and then cannot produce is worth
// less than one it declines to accept — see the interlace refusal below, which
// is the case that will actually turn up.
package config

import (
	"fmt"
	"math"
	"strconv"
	"strings"
)

// VideoFormatExample is the canonical spelling, used in every message that has
// to show the operator what the field wants and in the Settings screen's hint.
// One string, so the form the app asks for cannot drift between the three
// places it is quoted.
const VideoFormatExample = "1920x1080p50"

// Video dimension and frame-rate bounds. They are sanity limits, not a
// statement about what any encoder supports: their whole job is to turn a
// mistyped number into a message about this field instead of a caps string
// nothing can negotiate.
//
// 8192 covers 8K (7680x4320) with room above it. 1000 fps is far beyond any
// broadcast format and still refuses the kind of value a slipped keypress
// produces.
const (
	maxVideoDimension = 8192
	maxVideoFrameRate = 1000
)

// VideoFormatSpec is a parsed videoFormatOverride: the conform target for the
// contribution feed's video leg.
//
// It is NOT called VideoFormat, and the difference matters at the one place the
// two types meet. internal/m2lx.VideoFormat is what the SWITCHER REPORTS a node
// is currently carrying — a measurement, possibly of a format this application
// has never heard of, which is why it keeps Raw. This is what the operator has
// DECLARED the switcher is configured for. The derive-or-override decision has
// one of each in scope, and two types called VideoFormat in that function would
// be an invitation to assign the wrong one.
//
// Raw is kept for the same reason m2lx.VideoFormat keeps its own: it is what
// the operator typed, and it is what an error message, a log line or a Settings
// field should show them. Canonical() is what to show when the app is stating
// the format itself.
type VideoFormatSpec struct {
	// Raw is the value as it was written, trimmed of surrounding space and
	// nothing else. "1920X1080P50" stays upper-case here and normalises only in
	// Canonical.
	Raw string

	// Width and Height in pixels, both at least 1.
	Width  int
	Height int

	// FrameRateNum and FrameRateDen are the frame rate as an EXACT fraction:
	// 50/1, or 60000/1001 for the NTSC family. Den is never zero on a spec that
	// came out of ParseVideoFormat.
	//
	// A fraction rather than a float because that is what the format actually
	// is and what GStreamer's framerate field takes. 59.94 is not a frame rate;
	// it is a two-decimal rounding of 60000/1001, and a pipeline built from the
	// rounding is a pipeline whose timestamps drift against the switcher's by a
	// part in ten thousand for the length of a match.
	FrameRateNum int
	FrameRateDen int
}

// ParseVideoFormat parses "1920x1080p50" — <width>x<height>p<frame rate> — into
// a VideoFormatSpec. It is the ONLY parser of this grammar; nothing else in the
// application may take this string apart.
//
// The comparison is case-insensitive ("1920X1080P50" is accepted, because a
// capital X is what a keyboard produces as often as not) and surrounding space
// is ignored. Everything else is refused, and every refusal names the value and
// says what the field wants — this function's errors are read by an operator on
// the Settings screen, not by a developer in a log.
//
// FRAME RATES. Whole numbers are exact: p50 is 50/1. The NTSC family is
// accepted as the decimals people actually write and converted to the fraction
// it really is — 59.94 → 60000/1001, 29.97 → 30000/1001, 23.98 and 23.976 →
// 24000/1001 — because a broadcaster writes 59.94 and means 60000/1001, and a
// field that refused the spelling everyone uses would be a field nobody could
// fill in. Any other decimal is refused rather than rounded: 50.5 is not a
// format, it is a typo, and quietly conforming to 50.5/1 would be this
// function inventing a video standard.
func ParseVideoFormat(raw string) (VideoFormatSpec, error) {
	s := strings.TrimSpace(raw)
	if s == "" {
		// Reachable only from a caller that did not check for empty first —
		// Config.VideoFormatOverrideSpec and Validate both do, because empty is
		// the DEFAULT and means "derive from the switcher", not "wrong".
		return VideoFormatSpec{}, fmt.Errorf(
			"is empty; write it as %q, or leave it blank to take the format from the switcher",
			VideoFormatExample)
	}

	lower := strings.ToLower(s)
	if strings.ContainsAny(lower, " \t") {
		return VideoFormatSpec{}, fmt.Errorf("%q has spaces in it; write it as one word, e.g. %q", s, VideoFormatExample)
	}

	x := strings.IndexByte(lower, 'x')
	if x < 0 {
		return VideoFormatSpec{}, fmt.Errorf(
			"%q is not a video format; write it as <width>x<height>p<frame rate>, e.g. %q", s, VideoFormatExample)
	}

	rest := lower[x+1:]
	scan := strings.IndexAny(rest, "pi")
	if scan < 0 {
		return VideoFormatSpec{}, fmt.Errorf(
			"%q does not say the frame rate; write it as <width>x<height>p<frame rate>, e.g. %q", s, VideoFormatExample)
	}

	// INTERLACE IS REFUSED, AND THAT IS THE POINT OF READING THE LETTER AT ALL.
	//
	// 1080i25 is a real M2L-X configuration and somebody will type it. This
	// application's video leg produces progressive frames and there is no
	// interlacer in either bundler's element list, so accepting the string would
	// buy nothing but a later failure with a worse message. The refusal names
	// the letter, so an operator whose switcher really is interlaced learns that
	// this is a limitation of the app rather than a rejected spelling.
	if rest[scan] == 'i' {
		return VideoFormatSpec{}, fmt.Errorf(
			"%q is interlaced (the \"i\"); this application can only send progressive video, so an "+
				"interlaced switcher cannot be conformed to from here — write a progressive format such as %q "+
				"if the switcher can be set to one", s, VideoFormatExample)
	}

	width, err := parseVideoDimension(lower[:x], "width", s)
	if err != nil {
		return VideoFormatSpec{}, err
	}
	height, err := parseVideoDimension(rest[:scan], "height", s)
	if err != nil {
		return VideoFormatSpec{}, err
	}
	num, den, err := parseVideoFrameRate(rest[scan+1:], s)
	if err != nil {
		return VideoFormatSpec{}, err
	}

	return VideoFormatSpec{
		Raw:          s,
		Width:        width,
		Height:       height,
		FrameRateNum: num,
		FrameRateDen: den,
	}, nil
}

// parseVideoDimension reads the width or the height. what names which one, and
// whole names the value being parsed, so a message can quote both the part that
// is wrong and the format it came out of.
func parseVideoDimension(part, what, whole string) (int, error) {
	if part == "" {
		return 0, fmt.Errorf("%q has no %s; write it as <width>x<height>p<frame rate>, e.g. %q", whole, what, VideoFormatExample)
	}
	if !isAllDigits(part) {
		return 0, fmt.Errorf("%q has a %s of %q, which is not a whole number of pixels; e.g. %q",
			whole, what, part, VideoFormatExample)
	}
	// Atoi on an all-digit string fails only on overflow, which the bound below
	// would refuse anyway; it is checked rather than ignored so a 30-digit
	// paste cannot arrive as a negative.
	v, err := strconv.Atoi(part)
	if err != nil || v < 1 || v > maxVideoDimension {
		return 0, fmt.Errorf("%q has a %s of %q; it must be between 1 and %d pixels, e.g. %q",
			whole, what, part, maxVideoDimension, VideoFormatExample)
	}
	return v, nil
}

// parseVideoFrameRate reads the part after the "p" into an exact fraction. See
// ParseVideoFormat's header for which spellings are accepted and why the NTSC
// decimals are among them.
func parseVideoFrameRate(part, whole string) (num, den int, err error) {
	if part == "" {
		return 0, 0, fmt.Errorf("%q has no frame rate after the \"p\"; e.g. %q", whole, VideoFormatExample)
	}

	if isAllDigits(part) {
		v, err := strconv.Atoi(part)
		if err != nil || v < 1 || v > maxVideoFrameRate {
			return 0, 0, fmt.Errorf("%q has a frame rate of %q; it must be between 1 and %d, e.g. %q",
				whole, part, maxVideoFrameRate, VideoFormatExample)
		}
		return v, 1, nil
	}

	// The NTSC family. A decimal is accepted ONLY when it is the rounding of
	// some n*1000/1001 — which is what 23.98, 23.976, 29.97 and 59.94 all are —
	// and the exact fraction is what gets stored. The tolerance is 0.005, which
	// is wide enough for the two-decimal spellings people write (23.98 is 0.004
	// away from 24000/1001) and narrow enough that no rate anybody means as
	// itself falls inside it.
	f, ferr := strconv.ParseFloat(part, 64)
	if ferr != nil || math.IsNaN(f) || math.IsInf(f, 0) || f <= 0 || f > maxVideoFrameRate {
		return 0, 0, fmt.Errorf("%q has a frame rate of %q, which is not a number of frames per second; e.g. %q",
			whole, part, VideoFormatExample)
	}
	n := int(math.Round(f * 1001 / 1000))
	if n >= 1 && n <= maxVideoFrameRate && math.Abs(f-float64(n)*1000/1001) <= 0.005 {
		return n * 1000, 1001, nil
	}
	return 0, 0, fmt.Errorf(
		"%q has a frame rate of %q; whole numbers (24, 25, 30, 50, 60) and the NTSC rates "+
			"(23.98, 29.97, 59.94) are the frame rates there are, e.g. %q", whole, part, VideoFormatExample)
}

// isAllDigits reports whether every byte is an ASCII digit — and that the
// string is not empty, so a caller cannot read "" as a number.
//
// It is used instead of leaving the job to strconv because strconv.Atoi
// accepts a leading sign and an underscore separator, and "+50" or "5_0" as a
// frame rate is a value that parses and means the operator typed something
// else.
func isAllDigits(s string) bool {
	if s == "" {
		return false
	}
	for i := 0; i < len(s); i++ {
		if s[i] < '0' || s[i] > '9' {
			return false
		}
	}
	return true
}

// FrameRate is the frame rate as a float, for a message or a comparison. It is
// NOT what a pipeline should be built from — use FrameRateNum/FrameRateDen, or
// FrameRateFraction — because 60000/1001 has no exact float.
//
// Zero on the zero value, rather than a division by zero: a VideoFormatSpec{}
// is what a caller holds when there is no override, and the zero value must be
// safe to ask questions of.
func (f VideoFormatSpec) FrameRate() float64 {
	if f.FrameRateDen == 0 {
		return 0
	}
	return float64(f.FrameRateNum) / float64(f.FrameRateDen)
}

// FrameRateFraction renders the frame rate as "50/1" or "60000/1001": the form
// GStreamer's framerate field takes. It is a formatter and not a caps builder —
// assembling caps belongs to internal/gst, which is the only package that knows
// what else has to be in them.
func (f VideoFormatSpec) FrameRateFraction() string {
	return strconv.Itoa(f.FrameRateNum) + "/" + strconv.Itoa(f.FrameRateDen)
}

// Canonical renders the spec in the grammar's own spelling: "1920x1080p50",
// "1280x720p59.94". This is what to print when the APPLICATION is stating the
// format; Raw is what to print when quoting the operator back to themselves.
func (f VideoFormatSpec) Canonical() string {
	return fmt.Sprintf("%dx%dp%s", f.Width, f.Height, f.frameRateText())
}

// String makes a spec printable with %v and %s. It is Canonical, not Raw:
// anything that formats a spec without asking is the application talking, and
// the application should use its own spelling.
func (f VideoFormatSpec) String() string {
	return f.Canonical()
}

// frameRateText is the frame rate as it appears in Canonical: a whole number
// where the denominator is 1, and otherwise the familiar two- or three-decimal
// NTSC spelling with the trailing zeros taken off (59.94, not 59.940).
func (f VideoFormatSpec) frameRateText() string {
	if f.FrameRateDen == 1 {
		return strconv.Itoa(f.FrameRateNum)
	}
	s := strconv.FormatFloat(f.FrameRate(), 'f', 3, 64)
	s = strings.TrimRight(s, "0")
	return strings.TrimSuffix(s, ".")
}

// THERE IS DELIBERATELY NO ErrNoVideoFormat SENTINEL, and this note is here so
// that nobody adds one. "No override is set" is not an error condition — it is
// the DEFAULT, and it means derive the format from the switcher.
// Config.VideoFormatOverrideSpec reports it as (spec, false, nil), a shape a
// caller cannot mistake for a failure. A sentinel would invite exactly that
// mistake: every caller receiving one would have to remember it was not a
// problem, and the first one that forgot would refuse to start a feed over a
// field the operator had correctly left blank.

File: frontend/src/ui/videoformat.js
/**
 * THE VIDEO FORMAT THIS SEAT PRODUCES, and the switcher's own format beside it.
 *
 * Owner: WP-5b. Pure — no DOM, no browser API, no backend — so `node --test`
 * drives every rule below for real with nothing installed (videoformat.test.js).
 * settings.js is the wiring; this file is what it wires.
 *
 * ===================== WHY THIS IS A SELECTOR AND NOT A BOX =================
 *
 * videoFormatOverride was a free-text field. The operator's words: "the video
 * format should be a selector, not a free text field and it should show the
 * M2LX format clearly so it is obvious to the users when you diverge". Both
 * halves are failures of the same kind — a raster is not something anybody
 * should be typing from memory twenty minutes before kick-off, and a raster
 * typed correctly but DIFFERENTLY from the switcher's is refused by M2L-X with
 * nothing on this screen to have warned them.
 *
 * ===================== THE STRING STAYS A STRING ============================
 *
 * The config field is still the string "1920x1080p50" and that is deliberate,
 * argued at length in CONTRACT.md: a struct merges field-by-field, so a preset
 * carrying {"width":1280} would leave height at 1080 and conform the feed to
 * 1280x1080 — a raster nobody chose and no switcher accepts. A STRING CANNOT
 * HALF-ARRIVE. This module changes what the operator is shown; it changes
 * nothing about what is stored, and internal/config.ParseVideoFormat is still
 * the authority on what parses.
 *
 * So every value below is a CANONICAL SPELLING that both validate.js and
 * ParseVideoFormat accept, and videoformat.test.js runs the whole list through
 * validateConfig to prove it — an option an operator can select but not save
 * would be worse than the free-text box it replaced.
 *
 * ===================== EMPTY IS THE ANSWER, NOT AN ABSENCE ==================
 *
 * The default is FOLLOW_SWITCHER — the empty string — and it is the right
 * answer on almost every seat: the format is derived from the switcher at
 * START, which is the same derivation every position has always used. An
 * override is the exception, wanted only when the app has nothing to derive
 * from. That is why the follow option is kept OUT of the raster groups and
 * handed to the caller separately: it is not one entry among seventeen, it is
 * the thing the other seventeen are exceptions to.
 */

/**
 * FOLLOW_SWITCHER is the empty override: derive the format from the switcher.
 * It is a named constant because '' reads as "unset" everywhere else in this
 * application and here it is a decision.
 */
export const FOLLOW_SWITCHER = '';

/** The label on the follow option. It says what happens, not what is absent. */
export const FOLLOW_SWITCHER_LABEL = 'Follow the switcher (normal)';

/**
 * The rasters offered, grouped by picture size.
 *
 * WHY THESE AND NOT MORE. The list is the broadcast rasters an M2L-X deployment
 * is actually configured for; a longer list is a longer thing to scan and every
 * entry past the one the operator wants is a cost. Anything genuinely missing
 * still reaches config.json — a value already saved survives (planVideoFormats
 * keeps it), and the field remains a string, so nothing here can make a raster
 * unreachable for ever.
 *
 * The NTSC family is written as the decimals broadcasters write (59.94, not
 * 60000/1001) because that is what ParseVideoFormat accepts and what M2L-X
 * reports; internal/config/videoformat.go's parseVideoFrameRate resolves them
 * to the exact fractions within a 0.005 tolerance.
 */
export const VIDEO_FORMAT_GROUPS = Object.freeze([
  Object.freeze({
    label: '720',
    formats: Object.freeze(['1280x720p50', '1280x720p59.94', '1280x720p60']),
  }),
  Object.freeze({
    label: '1080',
    formats: Object.freeze([
      '1920x1080p23.98',
      '1920x1080p24',
      '1920x1080p25',
      '1920x1080p29.97',
      '1920x1080p30',
      '1920x1080p50',
      '1920x1080p59.94',
      '1920x1080p60',
    ]),
  }),
  Object.freeze({
    label: '2160 (UHD)',
    formats: Object.freeze([
      '3840x2160p25',
      '3840x2160p29.97',
      '3840x2160p30',
      '3840x2160p50',
      '3840x2160p59.94',
      '3840x2160p60',
    ]),
  }),
]);

/**
 * The provenance strings app.go stamps into ConformTargetView.Source. They are
 * the contract; videoformat.test.js pins them against app.go's own constants so
 * the two sides cannot drift into a state where every answer reads as unknown.
 */
export const CONFORM_SOURCE_SESSION = 'session';
export const CONFORM_SOURCE_OVERRIDE = 'override';
/**
 * The INSTANCE'S OWN SETTING, read live over REST by App.GetSwitcherFormat and
 * needing no session. GetConformTarget never stamps it — app.go says so in
 * terms — so a target carrying this source is always the independent number,
 * and that is what makes it usable here.
 */
export const CONFORM_SOURCE_SWITCHER = 'switcher';

/**
 * isIndependentOfTheOperator reports whether a target's raster came from
 * anywhere other than the operator's own videoFormatOverride.
 *
 * ===================== THE ONE RULE BOTH READOUTS OBEY ======================
 *
 * Both functions below exist to let an operator see that their override
 * disagrees with the facility. That comparison is worth nothing unless the
 * number on the other side of it is INDEPENDENT of the override, so this is the
 * single gate both go through, and it is written as an allowlist of the two
 * provenances that qualify rather than as "not override".
 *
 * An allowlist, because the failure modes are not symmetrical. A new provenance
 * that this build does not know about reads as NOT independent, so the readout
 * falls silent — the operator is told nothing, which is recoverable. A denylist
 * would read the same unknown provenance as independent and quote its raster as
 * the switcher's, which is the screen inventing a confirmation and is the one
 * thing a divergence warning must never do.
 *
 * @param {{source?: unknown}|null|undefined} target
 * @returns {boolean}
 */
export function isIndependentOfTheOperator(target) {
  if (!target || typeof target !== 'object') return false;
  return target.source === CONFORM_SOURCE_SESSION || target.source === CONFORM_SOURCE_SWITCHER;
}

/**
 * planVideoFormats decides what the <select> holds and which entry is selected.
 *
 * ===================== A SAVED VALUE IS NEVER DROPPED =======================
 *
 * `saved` may be a raster this list does not carry: a hand-edited config.json, a
 * preset from a facility running something unusual, or simply a format added to
 * Go's parser before it was added here. It is kept, in a group of its own that
 * says where it came from, because the alternative is a <select> that shows
 * 1280x720p50 while config.json says otherwise — the screen and the file
 * disagreeing, which is the fault describeDeviceSelection exists to prevent on
 * the device dropdowns and is no better here.
 *
 * @param {unknown} saved the stored videoFormatOverride
 * @returns {{follow: {value: string, label: string},
 *            groups: Array<{label: string, options: Array<{value: string, label: string}>}>,
 *            value: string}}
 */
export function planVideoFormats(saved) {
  const value = typeof saved === 'string' ? saved.trim() : '';

  const groups = VIDEO_FORMAT_GROUPS.map((g) => ({
    label: g.label,
    options: g.formats.map((f) => ({ value: f, label: f })),
  }));

  const known = value === FOLLOW_SWITCHER || VIDEO_FORMAT_GROUPS.some((g) => g.formats.includes(value));
  if (!known) {
    groups.push({
      label: 'Saved on this machine',
      options: [{ value, label: value }],
    });
  }

  return {
    follow: { value: FOLLOW_SWITCHER, label: FOLLOW_SWITCHER_LABEL },
    groups,
    value,
  };
}

/**
 * formatConformTarget renders a ConformTargetView as the canonical raster
 * string, or '' when there is nothing usable to render.
 *
 * The frame rate crosses the Wails boundary as the OPERATOR-FACING DECIMAL
 * already — app.go's ConformTargetView.FrameRate is gst.ConformTarget's
 * DisplayFrameRate, so 30000/1001 arrives as 29.97 — and String(Number(x)) is
 * what turns 50.0 back into "50" without turning 59.94 into "59.9".
 *
 * @param {{width?: unknown, height?: unknown, frameRate?: unknown}|null|undefined} target
 * @returns {string}
 */
export function formatConformTarget(target) {
  if (!target || typeof target !== 'object') return '';
  const w = target.width;
  const h = target.height;
  const r = target.frameRate;
  const positive = (n) => typeof n === 'number' && Number.isFinite(n) && n > 0;
  if (!positive(w) || !positive(h) || !positive(r)) return '';
  return `${w}x${h}p${String(Number(r))}`;
}

/**
 * describeConformTarget is what the operator is shown about the SWITCHER,
 * beside the control that can disagree with it.
 *
 * ===================== WHAT THE BINDING CAN AND CANNOT ANSWER ===============
 *
 * TWO BINDINGS CAN ANSWER THIS, and this function takes whichever the caller
 * has. App.GetSwitcherFormat reads the INSTANCE'S OWN SETTING over one REST call
 * and needs no session, which is what makes the readout work on a Settings
 * screen opened an hour before kick-off — the state the operator is actually in
 * when they are choosing a format. App.GetConformTarget answers from the RUNNING
 * session when there is one, and that answer is the switcher's too, because
 * Start derived it from the switcher.
 *
 * WHAT IS REFUSED is GetConformTarget's no-session answer, which is the
 * operator's own videoFormatOverride handed back stamped source="override".
 * Echoing an operator's own setting back to them under the heading "what M2L-X
 * is configured for" would be the screen inventing a confirmation, which is
 * worse than saying nothing: the whole point of the readout is to be the
 * INDEPENDENT number the override can be wrong against. That is why this
 * function reads `source` at all, and why the gate is
 * isIndependentOfTheOperator rather than a truthiness check on the raster.
 *
 * THE NOT-KNOWN LINE says the instance could not be read rather than "press
 * START", because pressing START is no longer what fixes it: with
 * GetSwitcherFormat wired, the remaining reasons are no host, not signed in yet,
 * or an instance that is not up — none of which START addresses, and all of
 * which resolve on their own once the seat is configured and reachable.
 *
 * @param {{source?: unknown}|null|undefined} target
 * @returns {string}
 */
export function describeConformTarget(target) {
  const raster = formatConformTarget(target);
  if (raster !== '' && isIndependentOfTheOperator(target)) {
    return `Switcher: ${raster}`;
  }
  return 'Switcher: not reachable yet';
}

/**
 * The four states of "does the override agree with the switcher".
 *
 *   follow    no override: the format is derived, which is right almost always
 *   unknown   an override is set and there is nothing independent to check it
 *             against — see describeConformTarget for why that is the state
 *             before START rather than a failure
 *   match     an override is set and it agrees with the running feed's target
 *   diverge   an override is set and it does NOT agree, which means the
 *             switcher will refuse this feed
 */
export const FORMAT_MATCH = Object.freeze({
  FOLLOW: 'follow',
  UNKNOWN: 'unknown',
  MATCH: 'match',
  DIVERGE: 'diverge',
});

/**
 * deriveFormatMatch compares the override on the form against the switcher's
 * own format and says, in one word and one short line, what that means.
 *
 * ===================== DIVERGENCE IS THE WHOLE POINT ========================
 *
 * Every source feeding an M2L-X instance must be produced in the instance's
 * format; one that is not is refused. So an override that disagrees is not a
 * preference, it is a feed that will not be accepted — and the operator has to
 * be able to see that WITHOUT knowing the rule, which is why the caller marks
 * the row as well as printing the line. The line is short on purpose: a
 * paragraph explaining conform rules is for whoever edits the code, and this
 * one has to be readable at a glance by somebody looking for a microphone.
 *
 * @param {unknown} override the videoFormatOverride on the form
 * @param {{source?: unknown}|null|undefined} target App.GetSwitcherFormat's
 *   answer, or App.GetConformTarget's when a session is running
 * @returns {{state: string, line: string, diverges: boolean}}
 */
export function deriveFormatMatch(override, target) {
  const value = typeof override === 'string' ? override.trim() : '';
  const raster = formatConformTarget(target);
  const independent = raster !== '' && isIndependentOfTheOperator(target);

  if (value === FOLLOW_SWITCHER) {
    return { state: FORMAT_MATCH.FOLLOW, line: '', diverges: false };
  }
  if (!independent) {
    return {
      state: FORMAT_MATCH.UNKNOWN,
      line: 'Overriding — the switcher’s format could not be read.',
      diverges: false,
    };
  }
  if (value === raster) {
    return { state: FORMAT_MATCH.MATCH, line: `Matches the switcher (${raster}).`, diverges: false };
  }
  return {
    state: FORMAT_MATCH.DIVERGE,
    line: `DIVERGES: the switcher is ${raster}. This feed will not be accepted.`,
    diverges: true,
  };
}
